index.js

```diff
@@ -22,14 +21,22 @@ const exec = (cmd, args) => {
 	).filter(Boolean);
 };
 
+const install = (bin, useYarn) => () => execa(bin, ['install'])
```

Comment:
Not sure if that's possible at the moment. The only thing you could do is work with the [skipped](https://github.com/SamVerschueren/listr#skipping-tasks) functionality. The downside of this is that it decides upfront if the task is skipped or not. So if you're in the `task` function, it's not possible to skip it afterwards because something failed.

What you could do however is running `yarn --version` in the `skip` function and detect if that works. If not, you can show a skipped message. So that can be solved this way, something like this.

```js
{
	title: 'Installing dependencies using Yarn',
	skip: () => execa('yarn', '--version')
		.then(() => false)
		.catch(() => 'Yarn not installed, falling back to npm')
	},
	task: () => execa('yarn')
}
```

About adding a new task dynamically, never tried it, I don't think it would work. What we can do for now is showing the same line but with `npm` and skip if the dependencies are already installed with Yarn.

### Solution 1

```js
{
	title: 'Installing dependencies using Yarn',
	skip: () => execa('yarn', '--version')
		.then(() => false)
		.catch(() => 'Yarn not installed, falling back to npm')
	},
	task: ctx => {
		ctx.yarn = true;
		return execa('yarn');
	}
},
{
	title: 'Installing dependencies using npm',
	skip: ctx => {
		if (ctx.yarn) {
			return 'Dependencies already installed with Yarn';
		}
	},
	task: () => execa('npm', ['install'])
}
```

But to be honest, that would look like crap. People might think we're going to install dependencies twice.

### Solution 2
Using [observables](https://github.com/SamVerschueren/listr#observable) to show a message

```js
{
	title: 'Installing dependencies',
	task: ctx => {
		return new Observable(observer => {
			execa('yarn')
				.then(() => observer.complete())
				.catch(() => {
					observer.next('Yarn not installed, falling back to npm');

					return execa('npm', ['install'])
						.then(() => observer.complete())
						.catch(err => observer.error(err));
				});
		});
	}
}
```

### Solution 3
Add something like a `visible` or `disabled` property to [Listr](https://github.com/SamVerschueren/listr) to dynamically show and hide tasks. That way we could use the `skip` function to detect if `yarn` is installed. If not, dynamically show the next task being `Installing dependencies via npm`.

-----------------------
The best option for now would be 2 I guess. But we will loose the `execa` output that is now generated because of `npm install`. So if we want to fix this on the long run, I will have to add new functionality to Listr.

Please provide me with feedback. If you have other solutions or needs that you want in order to solve this in the most elegant way possible, don't hesitate to throw it out. These are just the first things that come in my mind.